libyul/ObjectOptimizer.cpp

```diff
@@ -0,0 +1,169 @@
+			bool isCreation = !boost::ends_with(subObject->name, "_deployed");
+			optimize(
+				*subObject,
+				_settings,
+				isCreation
+			);
+		}
+
+	Dialect const& dialect = languageToDialect(_settings.language, _settings.evmVersion);
+	std::unique_ptr<GasMeter> meter;
+	if (EVMDialect const* evmDialect = dynamic_cast<EVMDialect const*>(&dialect))
+		meter = std::make_unique<GasMeter>(*evmDialect, _isCreation, _settings.expectedExecutionsPerDeployment);
+
+	std::optional<h256> cacheKey = calculateCacheKey(*_object.code, *_object.debugData, _settings, _isCreation);
+	if (cacheKey.has_value() && m_cachedObjects.count(*cacheKey) != 0)
+	{
+		overwriteWithOptimizedObject(*cacheKey, _object);
+		return;
+	}
+
+	OptimiserSuite::run(
+		dialect,
+		meter.get(),
+		_object,
+		_settings.optimizeStackAllocation,
+		_settings.yulOptimiserSteps,
+		_settings.yulOptimiserCleanupSteps,
+		_isCreation ? std::nullopt : std::make_optional(_settings.expectedExecutionsPerDeployment),
+		{}
+	);
+
+	if (cacheKey.has_value())
+		storeOptimizedObject(*cacheKey, _object, dialect);
+}
+
+void ObjectOptimizer::storeOptimizedObject(util::h256 _cacheKey, Object const& _optimizedObject, Dialect const& _dialect)
+{
+	m_cachedObjects[_cacheKey] = CachedObject{
+		std::make_shared<Block>(ASTCopier{}.translate(*_optimizedObject.code)),
+		&_dialect,
+	};
+}
+
+void ObjectOptimizer::overwriteWithOptimizedObject(util::h256 _cacheKey, Object& _object) const
+{
+	yulAssert(m_cachedObjects.count(_cacheKey) != 0);
+	CachedObject const& cachedObject = m_cachedObjects.at(_cacheKey);
+
+	yulAssert(cachedObject.optimizedAST);
+	_object.code = std::make_shared<Block>(ASTCopier{}.translate(*cachedObject.optimizedAST));
+	yulAssert(_object.code);
+
+	// There's no point in caching AnalysisInfo because it references AST nodes. It can't be shared
+	// by multiple ASTs and it's easier to recalculate it than properly clone it.
+	yulAssert(cachedObject.dialect);
+	_object.analysisInfo = std::make_shared<AsmAnalysisInfo>(
+		AsmAnalyzer::analyzeStrictAssertCorrect(
+			*cachedObject.dialect,
+			_object
+		)
+	);
+
+	// NOTE: Source name index is included in the key so it must be identical. No need to store and restore it.
+}
+
+std::optional<h256> ObjectOptimizer::calculateCacheKey(
+	Block const& _ast,
+	ObjectDebugData const& _debugData,
+	Settings const& _settings,
+	bool _isCreation
+)
+{
+	AsmPrinter asmPrinter(
+		AsmPrinter::TypePrinting::OmitDefault,
+		languageToDialect(_settings.language, _settings.evmVersion),
+		_debugData.sourceNames,
+		DebugInfoSelection::All()
+	);
+
+	bytes rawKey;
+	// NOTE: AsmPrinter never prints nativeLocations included in debug data, so ASTs differing only
+	// in that regard are considered equal here.  This is fine because the optimizer does not keep
+	// them up to date across AST transformations anyway so in any use where they need to be reliable,
+	// we just regenerate them by reparsing the object.
+	rawKey += keccak256(asmPrinter(_ast)).asBytes();
```

Comment:
Note that the way it is now, I am caching the object with wrong `nativeLocation`s, so when it's retrieved from the cache those are still be wrong. I expect that the user of the cache (i.e. `YulStack`) will reparse the source to correct the problem.

It would be a better to have it encapsulated here, so that the object is reparsed after it's optimized but before it's cached. Unfortunately to do this I'd have to duplicate the parsing and analyzing logic from `YulStack` here or do some refactor to share it so in the end I thought this was a better trade-off.